src/Parser/Core/Modules/RegenResourceCapTracker.js

```diff
@@ -0,0 +1,607 @@
+    const eventResource = this.getResource(event);
+    if (!eventResource) {
+      return;
+    }
+    const cost = this.getReducedCost(event);
+    if (!cost) {
+      // only interested in cast events that spend resource
+      return;
+    }
+    if (cost < 0) {
+      debug && console.warn(`${this.owner.formatTimestamp(event.timestamp, 3)} Unexpected negative cost ${cost} for spell ${event.ability.guid}`);
+    }
+    if (!this.constructor.exemptFromRefund.includes(event.ability.guid)) {
+      this.prevSpender = {
+        id: event.ability.guid,
+        cost,
+        timestamp: event.timestamp,
+      };
+    }
+    
+    const shouldAccumulate = eventResource.amount == null ||
+      this.constructor.castsToApplyCumulatively.includes(event.ability.guid) ||
+      this.isLastUpdateRecent(event.timestamp);
+    const current = (shouldAccumulate ? this.predictValue(event.timestamp) : eventResource.amount) - cost;
+    if (debug && !shouldAccumulate) {
+      this.debugActualVsPredicted(eventResource.amount, event.timestamp);
+    }
+    const max = this.currentMaxResourceWithFallback(event);
+    this.updateState(current, max);
+  }
+
+  on_byPlayer_damage(event) {
+    // only interested in damage events if they show a spending ability failing to connect (and so triggering a refund)
+    if (!this.prevSpender || event.ability.guid !== this.prevSpender.id ||
+        (event.timestamp - this.prevSpender.timestamp) > REFUND_SPENDER_WINDOW ||
+        event.tick || !HIT_TYPES_THAT_REFUND.includes(event.hitType)) {
+      return;
+    }
+    verboseDebug && console.log(`${this.owner.formatTimestamp(this.owner.currentTimestamp, 3)} attack didn't connect so restoring ${Math.round(this.constructor.resourceRefundOnMiss * 100)}% resource`);
+    const refund = Math.floor(this.prevSpender.cost * this.constructor.resourceRefundOnMiss);
+    const current = this.predictValue(event.timestamp) + refund;
+    this.updateState(current);
+  }
+
+  on_toPlayer_drain(event) {
+    if(event.resourceChangeType !== this.constructor.resourceType.id || !event.resourceChange) {
+      return;
+    }
+    const eventResource = this.getResource(event);
+    const drain = this.getReducedDrain(event);
+    
+    const shouldAccumulate = !eventResource || eventResource.amount == null ||
+      this.constructor.energizersToApplyCumulatively.includes(event.ability.guid) ||
+      this.isLastUpdateRecent(event.timestamp);
+    
+    // eventResource.amount for a drain is the value before the change
+    const current = (shouldAccumulate ? this.predictValue(event.timestamp) : eventResource.amount) - drain;
+    if (debug && !shouldAccumulate) {
+      this.debugActualVsPredicted(eventResource.amount, event.timestamp);
+    }
+    const max = this.currentMaxResourceWithFallback(event);
+    this.updateState(current, max);
+  }
+
+  isLastUpdateRecent(timestamp) {
+    return this.regenState.timestamp + this.constructor.cumulativeEventWindow >= timestamp;
```

Comment:
Using parens around multiple conditionals is recommended so the code is easier to understand